packages/@react-aria/toast/src/useToastRegion.ts

```diff
@@ -29,14 +29,79 @@ export function useToastRegion<T>(props: AriaToastRegionProps, state: ToastState
   let stringFormatter = useLocalizedStringFormatter(intlMessages, '@react-aria/toast');
   let {landmarkProps} = useLandmark({
     role: 'region',
-    'aria-label': props['aria-label'] || stringFormatter.format('notifications')
+    'aria-label': props['aria-label'] || stringFormatter.format('notifications', {count: state.visibleToasts.length})
   }, ref);
 
   let {hoverProps} = useHover({
     onHoverStart: state.pauseAll,
     onHoverEnd: state.resumeAll
   });
 
+  // Manage focus within the toast region.
+  // If a focused containing toast is removed, move focus to the next toast, or the previous toast if there is no next toast.
+  // We might be making an assumption with how this works if someone implements the priority queue differently, or
+  // if they only show one toast at a time.
+  let toasts = useRef([]);
+  let prevVisibleToasts = useRef(state.visibleToasts);
+  let focusedToast = useRef(null);
+  useLayoutEffect(() => {
+    // If no toast has focus, then don't do anything.
+    if (focusedToast.current === -1 || state.visibleToasts.length === 0) {
+      toasts.current = [];
+      prevVisibleToasts.current = state.visibleToasts;
+      return;
+    }
+    toasts.current = [...ref.current.querySelectorAll('[role="alertdialog"]')];
+    // If the visible toasts haven't changed, we don't need to do anything.
+    if (prevVisibleToasts.current.length === state.visibleToasts.length
+      && state.visibleToasts.every((t, i) => t.key === prevVisibleToasts.current[i].key)) {
+      prevVisibleToasts.current = state.visibleToasts;
+      return;
+    }
+    // Get a list of all removed toasts by index.
+    let removedToasts = prevVisibleToasts.current
+      .map((t, i) => ({
+        ...t,
+        i,
+        isRemoved: !state.visibleToasts.some(t2 => t.key === t2.key)
+      }));
+
+    let removedToastIndex = removedToasts.findIndex(t => t.i === focusedToast.current);
+
+    // If the focused toast was removed, focus the next or previous toast.
+    if (removedToastIndex > -1) {
+      let i = 0;
+      let nextToast;
+      let prevToast;
+      while (i <= removedToastIndex) {
+        if (!removedToasts[i].isRemoved) {
+          prevToast = Math.max(0, i - 1);
+        }
+        i++;
+      }
+      while (i < removedToasts.length) {
+        if (!removedToasts[i].isRemoved) {
+          nextToast = i - 1;
+          break;
+        }
+        i++;
+      }
```

Comment:
bit of a nit, but I found this a bit hard to follow, could we instead just slice the removedToasts list into two (one going from the 0 -> removedToastIndex and the other going from removedToastIndex -> end) and use findLastIndex with the same `!removedToasts.isRemoved` condition on the first slice to get the index of the prevToast and findIndex + the removedToastIndex on the second slice to get the nextToast index?

Thinking on this, it might just be the "removedToasts" naming that is tripping me up since that list contains all the toasts but just adds extra data to mark the removed ones.

Reply:
yeah, it was hard to write too
I've clarified with renames and some more comments